sycl/doc/design/MappingHostAddressesToDeviceEntities.md

```diff
@@ -0,0 +1,369 @@
+variable with the same name in the unnamed namespace.  The integration footer
+can reference the variable in the global namespace as `::FuBar`, but there is
+no way to reference the variable in the unnamed namespace using fully qualified
+lookup.
+
+Such programs are still legal, though.  The integration footer can support
+cases like this by defining a shim function that returns a reference to the
+shadowed device global:
+
+```
+namespace {
+namespace __sycl_detail {
+
+static constexpr decltype(FuBar) &__shim_1() {
+  return FuBar;   // References 'FuBar' in the unnamed namespace
+}
+
+} // namespace __sycl_detail
+} // namespace (unnamed)
+
+namespace sycl::detail {
+
+__sycl_device_global_registration::__sycl_device_global_registration() noexcept {
+  __register_uniquely_identifiable_object(
+    &::FuBar,
+    /* same string returned from __builtin_sycl_unique_stable_id(::FuBar) */,
+    "device_global");
+  __register_uniquely_identifiable_object(
+    &::__sycl_detail::__shim_1(),
+    /* same string returned from __builtin_sycl_unique_stable_id(::(unnamed)::FuBar) */,
+    "device_global");
+}
+
+} // namespace sycl::detail
+```
+
+The `__shim_1()` function is defined in the same namespace as the second
+`FuBar` device global, so it can reference the variable through unqualified
+name lookup.  Furthermore, the name of the shim function is globally unique, so
+it is guaranteed not to be shadowed by any other name in the translation unit.
+This problem with variable shadowing is also a problem for the integration
+footer we use for specialization constants.  See the [specialization constant
+design document][5] for more details on this topic.
+
+[5]: <SpecializationConstants.md>
+
+## Custom host compiler approach
+
+With this approach, we simply schedule one more pass in the optimization
+pipeline, which should be executed regardless of the optimization level, because
+it is required for proper functioning of some features.
+
+The pass has functionality similar to the integration footer, i.e. it emits a
+global constructor which in turn calls `__register_uniquely_identifiable_object`
+to provide the runtime with required mapping information.
+
+Unlike with the integration footer approach, no separate file is being
+generated. This preserves all source files mapping and checksums to be in place
+and correct.
+
+Generated constructor function should have internal linkage to avoid possible
+names clashes and multiple definition errors later at link stage.
+
+Generated constructor contains a call to
+`__register_uniquely_identifiable_object` for each global variable which has
+`sycl-unique-id` and `sycl-uid-kind` attributes, passing values of those
+attributes into the corresponding arguments of the function.
+
+### Handling shadowed variables
+
+Unlike with the integration footer the problem with shadowed variables doesn't
+really exists with the custom host compiler approach, because it is compiler
```

Comment:
```suggestion
really exists with the custom host compiler approach, because it is compiler's
```

Reply:
Applied in 96c0e37279fa1a74b4db818b59e85d98a50379a6